Add NSenter tunnel that allows to enter in any linux namespace container

amigrave opened the pull request:
NSEnter: A tunnel to a process on the same host, launched with nsenter.

This is a generic tunnel for all containers using linux namespaces such as docker, systemd-nspawn, rkt, ... that can be spawned as a subprocess of any given process id. This tunnel uses sudo hence it
requires the same passworless setup than the `Sudo` tunnel plus the ``nsenter`` command line utility found in ``linux-utils`` package.

amigrave committed:
Add NSenter and NSpawn tunnels

NSEnter allows to enter in any linux namespace container
NSpawn uses systemd-nspawn to create a container (docket alternative)

amigrave commented:
Hi @lordmauve , here's a pull request about an `nsenter` tunnel we discussed about at the Europython.

I also have a couple of questions about chopsticks, can you tell where's the best place to ask them ?

Thanks.

lordmauve commented:
Thanks, this is interesting. I'd like to review some documentation for `nsenter` before I merge this, if you can wait a few days.

There is a [Gitter for general questions about Chopsticks](https://gitter.im/chopsticks-chat/Lobby).

amigrave commented:
> Thanks, this is interesting. I'd like to review some documentation for nsenter before I merge this, if you can wait a few days.

of course !

> There is a Gitter for general questions about Chopsticks.

ok, see you there, then. I sent you a mail some time ago but it must have landed in /dev/null or spam :)

amigrave commented:
> Thanks, this is interesting. I'd like to review some documentation for nsenter before I merge this, if you can wait a few days.

I just realized that maybe I should have use `ctypes` in order to do the `setns` system calls from python instead of executing `nsenter`. That would need a bit of refactoring of `SubprocessTunnel` in order to make the `subprocess.Popen`'s `preexec_fn` callable overloadable but it's not a big deal. 

Would you mind keeping this PR "on hold" until I have time to make some tests ?

lordmauve commented:
I don't think we should maintain ctypes bindings for nsenter within Chopsticks. If you create a library we could make Chopsticks use it if it is installed.

amigrave commented:
You're right! There's already [python-nsenter](https://github.com/zalando/python-nsenter).

But now that I'm thinking about it, I realize that there's an impact for the enduser.
The `NSEnter` tunnel I currently proposed in this PR is working like the `Sudo` tunnel, so the user is supposed to setup it's passwordless sudo config in order to allow the NSEnter tunnel to use `sudo` when launching the `nsenter` command.

If we're making the system calls ourselves via `python-nsenter`, it means that the script using chopsticks should be launched with the root permission which might not be what a user wants and is probably not the usage of chopsticks you foreseen.

So before going further I'd like to have your opinion on that topic.

amigrave committed:
Allow to setup logging on remote process (through `stderr`)

amigrave committed:
Fix bug preventing decorated functions to be called remotely (Fixes #50)

amigrave committed:
Allow site-packages imports on remote via allow_site_imports

By design, chopsticks disables site-packages on the remote (-usS) to ensure
the master's code is always used. However, in some cases using the remote's
own site-packages is preferable: faster execution (no need to serve packages
over the wire), the remote environment is considered the reference, or to
avoid issues with packages relying on C extensions that chopsticks cannot
serve.

Set chopsticks.allow_site_imports = True or CHOPSTICKS_ALLOW_SITE_IMPORTS=1
to switch the remote Python invocation from -usS to -u. The setting propagates
through tunnel chains automatically.

amigrave committed:
Add `local_imports` context manager (allow to disable tunnelled imports)

Example:

    def remote_function():
        with chopsticks.tunnel.local_imports():
            import psycopg2
        ...

Icallhimtest committed:
Set running=False when IOLoop stops

Steps to reproduce:

1) Make a tunnel then close it - this ends the IOLoop as
   self.{read,write} are empty, but without calling `stop()`
2) Make another tunnel - stderr ioloop never starts running
   as it's still considered running https://github.com/amigrave/chopsticks/blob/master/chopsticks/tunnel.py#L51
   -> we never get the stderr

This error is hidden if the second tunnel is created before the first one is closed.

amigrave committed:
Bump version

lade-odoo committed:
Stop using deprecated (>= 3.12) `imp` module

As of Python 3.12, the `imp` module is no longer supported. Some other
behaviour are also deprecated, that is why `find_spec` has been added.

Even though, my use case works, we are not shielded from another
issue while using it.

amigrave committed:
Double the import timeout, from 5 to 10

Tunnelled imports with high latency can be slow

amigrave committed:
Add SSHTunnel 'host_check' nad `forward_agent` options

amigrave committed:
Improve context recurion and provide inception level

Inception level == how deep the rabbit hole is

amigrave committed:
Port is provided as int but should be converted as string for cmd

amigrave committed:
Fix LOAD_GLOBAL index for Python 3.11+

amigrave committed:
Bump version to 1.0.2

amigrave committed:
Add fallback to importlib when handle_imp can't find module in sys.path

Modern editable installs (pip >= 21.3 + pyproject.toml) register packages via
importlib.metadata rather than adding the source directory directly to sys.path.
The filesystem-only search in handle_imp would return exists=False for such
packages, preventing remotes from importing them (notably chopsticks itself).

Add an importlib.util.find_spec() fallback so any importable module on the
controller can be served to remotes regardless of how it was installed.

amigrave committed:
Handle 4-tuple from dis._unpack_opargs on Python 3.13

dis._unpack_opargs is a private API that changed its return tuple from
(offset, op, arg) to (start, end, op, arg) in Python 3.13. Use item[-2:]
instead of fixed unpacking to stay compatible across versions.

amigrave committed:
Return GroupResult immediately when all tunnels failed to connect

When _connect() returns an empty list (all tunnels refused), _parallel is
called with no tunnels, GroupOp.waiting stays at 0, and loop.stop is never
triggered. loop.run() then returns None instead of a GroupResult, causing
AttributeError on any subsequent .failures() call.

Return GroupResult(op.results) immediately when waiting == 0, which also
carries the connection errors already stored in op.results.

amigrave committed:
Add connect_timeout option and set default to 10 seconds

amigrave committed:
Encode bubble as base64 to avoid UnicodeDecodeError on relay sub-hosts

amigrave committed:
Fix some ioloop race conditions

- StderrReader.on_data: catch OSError (EBADF) when fd is closed between
  select() and read(), e.g. when a tunnel closes while the errloop thread
  is still running. Avoids unhandled exception in Thread-1.

- IOLoop.step xs loop: use getattr(reader, 'errback', None) before calling
  errback, since reader is a plain callback (bound method) that may not have
  an errback attribute. Avoids AttributeError on stale fds in errloop.

- IOLoop.step rs loop: use read.pop(r, None) with a None guard instead of
  read.pop(r)(), since a fd can appear in both rs (readable) and xs
  (exceptional) in the same select() call. The xs loop removes it from
  self.read first, causing KeyError in the rs loop. Avoids KeyError on
  concurrent readable+exceptional fds.

amigrave committed:
Bump version to 1.0.3

amigrave committed:
Add get_current_function() to inspect the active remote call

amigrave committed:
Fix import wakeups under concurrent loading

Wait until the requested module is actually cached instead of assuming any
loader wakeup means the import completed.

This avoids raising import timeouts when several threads are waiting on the
same loader condition.

amigrave committed:
Use poll() in the ioloop when available

select() hits platform fd limits under heavier tunnel activity, especially on
macOS where higher fd numbers can fail unexpectedly.

Use poll() when available and keep the existing select()-based path as a
fallback.

amigrave committed:
Make the tunnel IOLoop thread-local

The global loop is not re-entrant and breaks when multiple threads call into
the same tunnel at the same time.

Create one IOLoop per thread while keeping the existing module-level loop API.

amigrave committed:
Stop stderr readers cleanly before fd reuse

A stderr reader could remain registered while a tunnel is closing and race
with its own re-registration path.

Stop it before closing the subprocess pipes and synchronize shutdown so stale
callbacks cannot survive on recycled file descriptors.

amigrave committed:
Add keep-alive support to SSH tunnels

Long-running runs can leave SSH connections idle long enough for intermediate
network equipment to drop them.

Allow configuring SSH keep-alives so inactive tunnels stay up during slower or
uneven parallel workloads.

amigrave committed:
tunnel: read the bootstrap payload from the duplicated stdin

Reading the bootstrap payload from `sys.stdin` could consume extra buffered
bytes and shift the protocol stream seen by the tunnel pipe.

Read it from the duplicated stdin pipe instead so the bootstrap and protocol
traffic stay on the same buffered reader.

amigrave committed:
Bump version to 1.0.4
